server/src/frameworks/Hardhat/HardhatProject.ts

```diff
@@ -0,0 +1,298 @@
+          detached: true,
+          execArgv: [],
+        }
+      );
+      this.workerStatus = WorkerStatus.INITIALIZING;
+
+      this.workerProcess.on("message", async (message: Message) => {
+        await this._handleMessage(message);
+      });
+
+      this.workerProcess.on("error", (err) => {
+        this.logger.error(err);
+      });
+
+      this.workerProcess.on("exit", async (code) => {
+        this.logger.error(`Child process exited: ${code}`);
+        this.workerStatus = WorkerStatus.ERRORED;
+        this._onInitialized();
+      });
+    });
+  }
+
+  public async fileBelongs(sourceURI: string): Promise<boolean> {
+    const workerPromise = new Promise((resolve, reject) => {
+      this._checkWorkerExists();
+      this._checkWorkerNotInitializing();
+
+      if (this.workerStatus === WorkerStatus.RUNNING) {
+        // HRE was loaded successfully. Delegate to the worker that will use the configured sources path
+        const requestId = this._prepareRequest(resolve, reject);
+
+        this.workerProcess!.send(new FileBelongsRequest(requestId, sourceURI));
+      } else {
+        // HRE could not be loaded. Claim ownership of all solidity files under root folder
+        // This is to avoid potential hardhat-owned contracts being assigned to i.e. projectless
+        resolve(directoryContains(this.basePath, sourceURI));
+      }
+    });
+
+    return Promise.race([
+      workerPromise,
+      this._requestTimeout("fileBelongs"),
+    ]) as Promise<boolean>;
+  }
+
+  public async buildCompilation(
+    sourceUri: string,
+    openDocuments: OpenDocuments
+  ): Promise<CompilationDetails> {
+    const workerPromise = new Promise((resolve, reject) => {
+      this._checkWorkerExists();
+      this._checkWorkerNotInitializing();
+
+      if (this.workerStatus === WorkerStatus.ERRORED) {
+        throw new Error(this.workerLoadFailureReason);
+      }
+
+      const requestId = this._prepareRequest(resolve, reject);
+
+      this.workerProcess!.send(
+        new BuildCompilationRequest(requestId, sourceUri, openDocuments)
+      );
+    });
+
+    return Promise.race([
+      workerPromise,
+      this._requestTimeout("buildCompilation"),
+    ]) as Promise<CompilationDetails>;
+  }
+
+  private _prepareRequest(
+    resolve: (value: unknown) => void,
+    reject: (reason?: any) => void
+  ): number {
+    this.requestId++;
+
+    this._onResponse[this.requestId] = resolve;
+    this._onError[this.requestId] = reject;
+
+    return this.requestId;
+  }
+
+  public async onWatchedFilesChanges({
+    changes,
+  }: DidChangeWatchedFilesParams): Promise<void> {
+    for (const change of changes) {
+      if (change.uri.endsWith(".sol")) {
+        this.workerProcess?.send(new InvalidateBuildCacheMessage());
+      } else if (change.uri === this.configPath) {
+        this.logger.info(`Config file changed. Restarting worker process.`);
+
+        // Kill existing worker process
+        this.workerProcess?.kill("SIGKILL");
```

Comment:
Should we update the state here to stop any messages coming in during initialization?

Reply:
It happens naturally since the worker status is set to `initializing`. If some validation request happens to come at this point, it will fail with a descriptive error message